Automattic/wp-calypso
# [package.json] Enable max_old_space_size Override for Desktop

The behavior of heap limits and garbage collection in Node 12 has changed such that attempts to build wp-desktop with updated Calypso (which is now using Node 12) fail due to memory pressure. As-is, Node will attempt to use up to the set heap allocation (`max_old_space_size`) of 8GB but runs into the hard 4GB limit of Circle CI's RAM [at which point Node kills the build process](https://circleci.com/gh/Automattic/wp-desktop/50959). 

```
// ...
lerna success - photon
lerna success - @automattic/tree-select
Rendering Complete, saving .css file...
Wrote CSS to /home/circleci/wp-desktop/calypso/public/directly.css
Rendering Complete, saving .css file...
Wrote CSS to /home/circleci/wp-desktop/calypso/public/editor.css
Killed
```

We need to tell Node to allocate to the heap _less_  than the Docker container's 4GB RAM limit. However, it seems there's a fundamental flaw in the way `NODE_ARGS` is set such that this parameter cannot be overriden at build time.

#### Simplified Example

A simple example to illustrate:

```json
"scripts": {
  "run-my-script": "ENV_VAR=default npm run my-script",
  "my-script": "echo $ENV_VAR"
}
```

```
> ENV_VAR=custom npm run run-my-script
> echo $ENV_VAR
> default

// Want "custom", but get "default"
// Not what we intend or expect
```
In a nutshell, the above example illustrates the problem with the way `NODE_ARGS` is currently set. We need to be able to override this value when building wp-desktop, but are unable to.

(Side note: For some reason, this hasn't been an issue until Node 12 updated the way memory is managed. In addition, this only seems to affect Calypso as it is built for wp-desktop. However the issue with the way `NODE_ARGS` is set is a problem regardless.)

#### Changes proposed in this Pull Request
The values of `NODE_ARGS` and `max_old_space_size` for wp-desktop should be able to be overriden per the constraints of the hardware Caypso is being built on. Using the syntax `NODE_ARGS=${NODE_ARGS:---max_old_space_size=8192}` ensures that the default value of 8192MB will be used unless overridden by the caller, so we can maintain the existing behavior and customize this parameter where necessary.

#### Testing instructions

By pointing wp-desktop at this branch, we can verify that enabling the override of the hard-coded `max_old_space_size` value fixes the build failure on CI for wp-desktop.

- **Before**: out-of-memory - process killed ([build job](https://circleci.com/gh/Automattic/wp-desktop/50959)]
- **After**: `max_old_space_size` override is respected, and Calypso build is successful ([build job](https://circleci.com/gh/Automattic/wp-desktop/51034))
